TrackingTools/Producers/src/AnalyticalPropagatorESProducer.cc

```diff
@@ -13,38 +18,46 @@
 
 using namespace edm;
 
-AnalyticalPropagatorESProducer::AnalyticalPropagatorESProducer(const edm::ParameterSet& p) {
-  std::string myname = p.getParameter<std::string>("ComponentName");
-  pset_ = p;
-  setWhatProduced(this, myname);
+class AnalyticalPropagatorESProducer : public edm::ESProducer {
+public:
+  AnalyticalPropagatorESProducer(const edm::ParameterSet& p);
+  std::unique_ptr<Propagator> produce(const TrackingComponentsRecord&);
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& oDesc);
+
+private:
+  const edm::ESGetToken<MagneticField, IdealMagneticFieldRecord> magToken_;
+  const double dphiCut_;
+  const PropagationDirection dir_;
+};
+
+AnalyticalPropagatorESProducer::AnalyticalPropagatorESProducer(const edm::ParameterSet& p)
+    : magToken_{setWhatProduced(this, p.getParameter<std::string>("ComponentName"))
+                    .consumesFrom<MagneticField, IdealMagneticFieldRecord>(
+                        edm::ESInputTag("", p.getParameter<std::string>("SimpleMagneticField")))},
+      dphiCut_{p.getParameter<double>("MaxDPhi")},
+      dir_{[](std::string const& pdir) {
```

Comment:
given the number of places where this happens
https://cmssdt.cern.ch/lxr/search?%21v=CMSSW_11_0_X_2019-06-04-2300&_filestring=.*.cc&_advanced=1&_string=%22oppositeToMomentum%22&_casesensitive=1
most probably is better to factorize it in a function somewhere, even in "PropagationDirection.h" itself...